File: lab1-rust/src/bsp/rpi3/mini_uart.rs
use super::physical_memory::map::mmio::MINI_UART_START;
use crate::bsp::common::MMIODerefWrapper;
use tock_registers::{
    interfaces::{Readable, Writeable},
    register_bitfields, register_structs,
    registers::{ReadOnly, ReadWrite},
};

use aarch64_cpu::asm;

register_bitfields! {
    u32,
    AUX_ENABLE [
        MINI_UART_ENABLE OFFSET(0) NUMBITS(1) [
            Disabled = 0,
            Enabled = 1
        ]
    ],
    AUX_MU_IER [
        RX_INT_ENABLE OFFSET(1) NUMBITS(1) [
            Disabled = 0,
            Enabled = 1
        ],
        TX_INT_ENABLE OFFSET(0) NUMBITS(1) [
            Disabled = 0,
            Enabled = 1
        ]
    ],
    AUX_MU_IIR [
        INT_PENDING OFFSET(0) NUMBITS(1) [
            NoInterrupt = 0,
            Interrupt = 1
        ]
    ],
    AUX_MU_LCR [
        WLEN OFFSET(0) NUMBITS(2) [
            SevenBit = 0b00,
            EightBit = 0b11
        ],
    ],
    AUX_MU_MCR [
        RTS OFFSET(1) NUMBITS(1) [
            Inactive = 0,
            Active = 1
        ]
    ],
    AUX_MU_LSR [
        RX_READY OFFSET(0) NUMBITS(1) [
            Empty = 0,
            Data = 1
        ],
        TX_EMPTY OFFSET(5) NUMBITS(1) [
            Full = 0,
            Empty = 1
        ]
    ],
    AUX_MU_CNTL [
        RX_ENABLE OFFSET(1) NUMBITS(1) [
            Disabled = 0,
            Enabled = 1
        ],
        TX_ENABLE OFFSET(0) NUMBITS(1) [
            Disabled = 0,
            Enabled = 1
        ]
    ],
    AUX_MU_BAUD [
        BAUDRATE OFFSET(0) NUMBITS(16) [],
    ],
}

register_structs! {
    #[allow(non_snake_case)]
    pub RegisterBlock {
        (0x00 => _reserved1),
        (0x04 => AUX_ENABLE: ReadWrite<u32, AUX_ENABLE::Register>),
        (0x08 => _reserved2),
        (0x40 => AUX_MU_IO: ReadWrite<u32>),
        (0x44 => AUX_MU_IER: ReadWrite<u32, AUX_MU_IER::Register>),
        (0x48 => AUX_MU_IIR: ReadWrite<u32, AUX_MU_IIR::Register>),
        (0x4C => AUX_MU_LCR: ReadWrite<u32, AUX_MU_LCR::Register>),
        (0x50 => AUX_MU_MCR: ReadWrite<u32, AUX_MU_MCR::Register>),
        (0x54 => AUX_MU_LSR: ReadOnly<u32, AUX_MU_LSR::Register>),
        (0x58 => AUX_MU_MSR: ReadOnly<u32>),
        (0x5C => AUX_MU_SCRATCH: ReadWrite<u32>),
        (0x60 => AUX_MU_CNTL: ReadWrite<u32, AUX_MU_CNTL::Register>),
        (0x64 => AUX_MU_STAT: ReadOnly<u32>),
        (0x68 => AUX_MU_BAUD: ReadWrite<u32, AUX_MU_BAUD::Register>),
        (0x6C => @END),
    }
}

/// Abstraction for the associated MMIO registers.
type Registers = MMIODerefWrapper<RegisterBlock>;

pub struct MiniUart {
    registers: Registers,
}

impl MiniUart {
    pub const unsafe fn new() -> Self {
        Self {
            registers: Registers::new(MINI_UART_START),
        }
    }

    pub fn setup_mini_uart(&self) {
        // Enable the mini UART.
        self.registers
            .AUX_ENABLE
            .write(AUX_ENABLE::MINI_UART_ENABLE::Enabled);
        // Disable TX, RX during configuration
        self.registers
            .AUX_MU_CNTL
            .write(AUX_MU_CNTL::RX_ENABLE::Disabled + AUX_MU_CNTL::TX_ENABLE::Disabled);
        // Disable interrupt
        self.registers
            .AUX_MU_IER
            .write(AUX_MU_IER::RX_INT_ENABLE::Disabled + AUX_MU_IER::TX_INT_ENABLE::Disabled);
        // Set the data size to 8 bit
        self.registers.AUX_MU_LCR.write(AUX_MU_LCR::WLEN::EightBit);
        // Don't need auto flow control
        self.registers.AUX_MU_MCR.write(AUX_MU_MCR::RTS::Inactive);
        // Set baud rate to 115200
        self.registers
            .AUX_MU_BAUD
            .write(AUX_MU_BAUD::BAUDRATE.val(270));
        // No FIFO
        self.registers
            .AUX_MU_IIR
            .write(AUX_MU_IIR::INT_PENDING::NoInterrupt);
    }

    pub fn send(&self, c: u8) {
        while !self.registers.AUX_MU_LSR.is_set(AUX_MU_LSR::TX_EMPTY) {
            // asm::nop();
        } // when TX is not empty, wait
        self.registers.AUX_MU_IO.set(c as u32);
    }

    pub fn receive(&self) -> u8 {
        while !self.registers.AUX_MU_LSR.is_set(AUX_MU_LSR::RX_READY) {
            // asm::nop();
        } // when RX is not ready, wait
        self.registers.AUX_MU_IO.get() as u8
    }
}

File: lab01/miniuart.c
// miniuart.c
#include "miniuart.h"
#include "gpio.h"


void miniUARTInit() {
    register unsigned int r;

    /* initialize UART */
    *AUX_ENABLE |= 1;   // Enable mini UART
    *AUX_MU_CNTL = 0;    // Disable TX, RX during configuration
    *AUX_MU_IER = 0;     // Disable interrupt
    *AUX_MU_LCR = 3;     // Set the data size to 8 bit
    *AUX_MU_MCR = 0;     // Don't need auto flow control
    *AUX_MU_BAUD = 270;  // Set baud rate to 115200
    *AUX_MU_IIR = 6;     // No FIFO

    /* map UART1 to GPIO pins */
    // page 101
    r =* GPFSEL1;
    r &= ~((7 << 12) | (7 << 15)); // gpio14, gpio15 innitial ->p92
    r |= (2 << 12) | (2 << 15);    // alt5 ->p102
    *GPFSEL1 = r;
    *GPPUD = 0;            // enable pins 14 and 15
    r = 150; while(r--) { asm volatile("nop"); }
    *GPPUDCLK0 = (1 << 14) | (1 << 15);
    r = 150; while(r--) { asm volatile("nop"); }
    *GPPUDCLK0 = 0;        // flush GPIO setup
    *AUX_MU_CNTL = 3;      // enable Tx, Rx // bit1:transmitter bit0:receiver
}

char miniUARTRead(void) {
    char r;
    /* wait until something is in the buffer */
    do{asm volatile("nop");}while(!(*AUX_MU_LSR & 0x01)); // This bit is set if the receive FIFO holds at least 1 symbol. 
    /* read it and return */
    r = (char)(*AUX_MU_IO);
    /* convert carrige return to newline */
    return r == '\r'?'\n':r;
}

void miniUARTWrite(char c) {
    /* wait until we can send */
    do{asm volatile("nop");}while(!(*AUX_MU_LSR & 0x20)); // This bit is set if the transmit FIFO can accept at least one byte. 
    /* write the character to the buffer */
    *AUX_MU_IO = c;
}

void miniUARTWriteString(const char* s) {
    while(*s) {
        /* convert newline to carrige return + newline */
        if(*s == '\n')
            miniUARTWrite('\r');
        miniUARTWrite(*s++);
    }
}

void miniUARTWriteHex(unsigned int d) {
    unsigned int n;
    int c;
    for (c = 28; c >= 0; c -= 4) {
        // get highest tetrad
        n = (d >> c) & 0xF;
        // 0-9 => '0'-'9', 10-15 => 'A'-'F'
        n += n > 9 ? 0x37 : 0x30;
        miniUARTWrite(n);
    }
}
